Pradeep/lucene fst

pradeepgv42 opened the pull request:
## Description
Add FST index to support regex queries on text, this is pulling in FST construction and matching code from lucene.
On disk index sizes are smaller than the text with letting go off of features and only supporting regex queries. 
Works better for prefix queries and currently support is only added for dictionary encoded columns. 

On some of the columns I tested on reduction in on-disk size is close to 4-6 times. 
To enable FST index, column has to be dictionary encoded and with below config.
FST Index currently generates two files: FST index file (<column>.lucene.fst) and 
inverted index from dictId to documents (<column>.dict.lucene.fst)

"fieldConfigList":[
  {
     "name":"text_col_1",
     "encodingType":"DICTIONARY",
     "indexType":"FST"
  }
]

Commit:
Add FST Index using lucene lib to speedup regexp queries
this is alternative to text_match but consuming less
disk space but more involved and hence flexible.

Realtime matching is using lucene automaton based matching
to keep the results consistent

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java:
```diff
@@ -161,16 +211,30 @@ private void createTextIndexForColumn(ColumnMetadata columnMetadata)
         for (int docId = 0; docId < numDocs; docId++) {
           textIndexCreator.addDoc(rawIndexReader.getString(docId, chunkReaderContext), docId);
         }
+        // TODO(pradeep): Handle non-dictionary case.
```
Why are we changing this code?

pradeepgv42 replied:
Oops my bad, forgot to remove this.

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java:
```diff
@@ -161,16 +211,30 @@ private void createTextIndexForColumn(ColumnMetadata columnMetadata)
         // read the raw value from dictionary using dictId
         try (BaseImmutableDictionary dictionary = getDictionaryReader(columnMetadata)) {
+          for (int dictId = 0; dictId < dictionary.length(); dictId++) {
+            luceneFSTIndexCreator.addSortedDictIds(
```
This will break the existing functionality provided and (used in Li) for text index - phrase/positional searches.

pradeepgv42 replied:
Removed

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java:
```diff
@@ -134,6 +136,53 @@ private void checkUnsupportedOperationsForTextIndex(ColumnMetadata columnMetadat
     }
   }
 
+  private void createFSTIndexForColumn(ColumnMetadata columnMetadata) throws IOException {
```
This should be enough for FST index creation. We shouldn't change createTextIndexForColumn() method as it will break the functionality

siddharthteotia replied:
On the same note, we should model this as a different text index type. Check TextIndexType enum. Currently it has Lucene.
May be we can have LUCENE_FST as another type.

pradeepgv42 replied:
Yup, I am calling it just "FST" in the FieldConfig  (https://github.com/pradeepgv42/incubator-pinot/pull/2/files#diff-6d51e3f41e478650e105a6b5960bd4f6) 
or are you referring to something else?

siddharthteotia replied:
Yes, new index type in FieldConfig is good.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:
```diff
@@ -109,8 +109,17 @@ private BaseFilterOperator constructPhysicalOperator(FilterContext filter,
+                          ((TextMatchPredicate) predicate).getValue(),
+                          _numDocs);
+              } else {
+                  return new TextMatchFilterOperator(
```
If I understand correctly, the FST can only be used to solve the regex matching query. If that is the case, we should use it for `REGEXP_LIKE` instead of `TEXT_MATCH`

siddharthteotia replied:
+1

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/invertedindex/RealtimeFSTIndexReader.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public class RealtimeFSTIndexReader implements InvertedIndexReader<MutableRoaringBitmap> {
```
This implementation does not use FST, and is not thread safe.
Also, it always scans all values in the `_docIdMap`, which is not very efficient

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/text/LuceneFSTIndexCreator.java:
```diff
@@ -0,0 +1,135 @@
+
+    // TODO: Replace this with a memory efficient dict.
+    private final FSTBuilder _fstBuilder;
+    private final Object2IntOpenHashMap<String> _stringValueToIndexMap;
```
Why maintaining a separate map inside the creator instead of passing in a dictionary? This map will potentially consume lots of heap memory, and cause memory issue for segment generation.

Jackie-Jiang reviewed (commented):
Seems FST is used to match a regex to the dictIds that matches the regex. IMO a better way to plug it in is through the `RegexpLikePredicateEvaluator`, and after getting the dictIds, we can get the docIds by scanning the forward index or leveraging the inverted index.

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/LuceneFSTIndexReader.java:
```diff
@@ -0,0 +1,117 @@
+        this._dataBuffer = PinotDataBuffer.mapFile(
+                fstFile, true,0,
+                fstFile.length(),
+                ByteOrder.BIG_ENDIAN, "fstIndexFile");
```
LITTLE_ENDIAN?

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -324,7 +347,8 @@ public void indexRow(GenericRow row) {
           // store the docID -> dictID mapping in forward index
           forwardIndexCreator.putDictId(dictId);
-          DictionaryBasedInvertedIndexCreator invertedIndexCreator = _invertedIndexCreatorMap.get(columnName);
+          DictionaryBasedInvertedIndexCreator invertedIndexCreator =
```
you should rebase. these interfaces are no longer there

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java:
```diff
@@ -134,6 +136,53 @@ private void checkUnsupportedOperationsForTextIndex(ColumnMetadata columnMetadat
     }
   }
 
+  private void createFSTIndexForColumn(ColumnMetadata columnMetadata) throws IOException {
```
How is this being called? I don't see change in the caller

siddharthteotia reviewed (commented):
Need to add support for reload and tests in SegmentPreprocessorTest, LoaderTest etc. All these had tests for text index to ensure V1 -> V3 conversion, reload, directory creation etcare working fine.

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -84,6 +86,7 @@
   private Map<String, ForwardIndexCreator> _forwardIndexCreatorMap = new HashMap<>();
   private Map<String, DictionaryBasedInvertedIndexCreator> _invertedIndexCreatorMap = new HashMap<>();
   private Map<String, DictionaryBasedInvertedIndexCreator> _textIndexCreatorMap = new HashMap<>();
+  private Map<String, DictionaryBasedInvertedIndexCreator> _fstIndexCreatorMap = new HashMap<>();
```
You should rebase. I think these interfaces are no longer there. 
Also, we should not base this off InvertedIndexReader and InvertedIndexCreator interface. Now we have separate interface for TextIndexReader and TextIndexCreator. It was recently decoupled from InvertedIndex interface

Commit:
cleanups of logs, unused headers and code

jackjlli committed:
Create swagger dir and put api dir to swagger dir in pinot-controller resource (#6122)

* Add swagger dir and put api dir to swagger dir in pinot-controller resource

Co-authored-by: Jack Li(Analytics Engineering) <jlli@jlli-mn1.linkedin.biz>

Jackie-Jiang committed:
Implement off-heap bloom filter reader (#6118)

Add BloomFilterCreator and BloomFilterReader interface (similar to other indexes)
Re-implemented `OnHeapGuavaBloomFilterCreator` and `OffHeapGuavaBloomFilterReader`
Remove the size limit of the bloom filter
Add `BloomFilterConfig` for custom `fpp`

The new `OffHeapGuavaBloomFilterReader` is compatible with the existing Guava bloom filter storage format, so it can read the bloom filter generated from the existing code.

In order to config `fpp` for a bloom filter, use the key `bloomFilterConfigs` under the `tableIndexConfig` in the table config.
E.g.
{
  "tableIndexConfig": {
    "bloomFilterConfigs": {
      "col1": {
        "fpp": 0.01
      }
    },
    ...
  },
  ...
}

xiangfu0 committed:
Make transform functions support underscore in the query functions (#6125)

akshayrai committed:
[TE] add owners field into the subscription validation (#6128)

xiangfu0 committed:
Fix superset docker image build script (#5965)

Commit:
[TE] Fix Thirdeye Maven publish (#6097)

This PR is to fix issue of not publishing thirdeye-dashboard.

KKcorps committed:
Change Signature of Broker API in Controller (#6119)

* Add code for v2 broker api

* Change v1 methods to use v2 signature

* Refactor: Change class name from ControllerBrokerResponse to InstanceInfo

* Refactor: Remove unused constant and reuse functions

* Refactor: Remove class name from the methods called from same class

Co-authored-by: Kartik Khare <kharekartik@Kartiks-MacBook-Pro.local>

xiangfu0 committed:
Fix meetup stream for quickstart (#6140)

jackjlli committed:
Put hadoop dependencies to pinot-orc, pinot-parquet modules (#6088)

Co-authored-by: Jack Li(Analytics Engineering) <jlli@jlli-mn1.linkedin.biz>

deemoliu committed:
add max length support in schema builder (#6112)

Add one more API to set maxLength in schemaBuilder. #6106

yupeng9 committed:
Adding the upsert support to real-time ingestion and query (#6113)

Part of a series of PRs for #4261

Added the major pieces for enabling upsert:
- the data structures for upsert metadata per table
- the upsert metadata processing during mutable segment ingestion
- the upsert metadata loading during segment loading
- enabling upsert's valid document filtering during query by enhancing the filter plan node
- the upsert metadata removal during segment removal
- an example quickstart using MeetupEventRSVP

Some more followup PRs will be sent out to improve the test coverage and feature polishing such as table config validation

npawar committed:
Remove flaky test from SegmentReducerTest (#6133)

mayankshriv committed:
Support for multi-threaded Group By reducer for SQL. (#6044)

The existing implementation of Broker reduce phase is single-threaded.
For group-by queries where large response are being sent back from multiple servers,
this could become a bottlenect.

Given that brokers are generally light on CPU usage, making the reduce phase
multi-threaded would be a good idea to boost performance. This PR adds a multi-threaded
implementation for the Group-By reducer for SQL.

- Added an executor service in BrokerReduceService that can be used by multi-threaded
  execution of the broker reduce phase. This is initialized with number of threads as:
  `Runtime.getRuntime().availableProcessors()`.

- Added a broker side config to specify max number of threads per query to be used for reduce phase.
  `pinot.broker.max.reduce.threads.per.query`. This has a same default value as server side combine phase:
  `Math.max(1, Math.min(10, Runtime.getRuntime().availableProcessors() / 2))`

   For reverting to single threaded reduce, set this config to 1.

- The GroupByDataTableReducer uses the following algorithm for determining number of
  threads to use in reduce phase (per query):
  - If there are less than 2 data tables to reduce, it uses single threaded run.
  - Else, it uses `Math.min(pinot.broker.max.reduce.threads.per.query, numDataTables).

- For testing, explicitly sets num threads to reduce to be > 1 to ensure functional
  correctness is tested.

Commit:
Remove preconditions check on bitmapbasedfilteroperation and return empty filter block when no matching dictionary items

Commit:
SortedIndexBasedFilter for sorted index in regexp_like

Commit:
Only enable bitmap/sortedindex when fst index is there and also set isAlwaysTrue/False for regexp evaluator

Commit:
remove unused imports

pradeepgv42 commented:
fyi, closing this, there's a PR against apache:master
